circuit/src/integer/mod.rs

```diff
@@ -250,100 +91,611 @@ where
+	IntegerReduceChip<W, N, NUM_LIMBS, NUM_BITS, P>
+where
+	P: RnsParams<W, N, NUM_LIMBS, NUM_BITS>,
+{
+	/// Creates a new reduce chip
+	pub fn new(assigned_integer: &AssignedInteger<W, N, NUM_LIMBS, NUM_BITS, P>) -> Self {
+		Self {
+			assigned_integer: assigned_integer.clone(),
+			_native: PhantomData,
+			_wrong: PhantomData,
+			_rns: PhantomData,
+		}
+	}
+}
 
-	/// Assign cells for reduce operation.
-	pub fn reduce(
-		x_limbs: [AssignedCell<N, N>; NUM_LIMBS],
-		rw: ReductionWitness<W, N, NUM_LIMBS, NUM_BITS, P>, config: IntegerConfig<NUM_LIMBS>,
-		mut layouter: impl Layouter<N>,
-	) -> Result<[AssignedCell<N, N>; NUM_LIMBS], Error> {
+impl<W: FieldExt, N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P> Chip<N>
+	for IntegerReduceChip<W, N, NUM_LIMBS, NUM_BITS, P>
+where
+	P: RnsParams<W, N, NUM_LIMBS, NUM_BITS>,
+{
+	type Output = AssignedInteger<W, N, NUM_LIMBS, NUM_BITS, P>;
+
+	/// Make the circuit config.
+	fn configure(common: &CommonConfig, meta: &mut ConstraintSystem<N>) -> Selector {
+		let selector = meta.selector();
+		let p_in_n = P::wrong_modulus_in_native_modulus();
+
+		meta.create_gate("reduce", |v_cells| {
+			let s = v_cells.query_selector(selector);
+			let mut t_exp = [(); NUM_LIMBS].map(|_| None);
+			let reduce_q_exp = v_cells.query_advice(common.advice[NUM_LIMBS], Rotation::prev());
+			let mut limbs_exp = [(); NUM_LIMBS].map(|_| None);
+			let mut residues_exp = Vec::new();
+			let mut result_exp = [(); NUM_LIMBS].map(|_| None);
+			for i in 0..NUM_LIMBS {
+				t_exp[i] = Some(v_cells.query_advice(common.advice[i], Rotation::prev()));
+				limbs_exp[i] = Some(v_cells.query_advice(common.advice[i], Rotation::cur()));
+				if i < NUM_LIMBS / 2 {
+					residues_exp.push(v_cells.query_advice(common.advice[i], Rotation::next()));
+				}
+				result_exp[i] =
+					Some(v_cells.query_advice(common.advice[i + NUM_LIMBS], Rotation::next()));
+			}
+			let t_exp = t_exp.map(|x| x.unwrap());
+			let limbs_exp = limbs_exp.map(|x| x.unwrap());
+			let result_exp = result_exp.map(|x| x.unwrap());
+
+			// REDUCTION CONSTRAINTS
+			let mut constraints =
+				P::constrain_binary_crt_exp(t_exp, result_exp.clone(), residues_exp);
+			// NATIVE CONSTRAINTS
+			let native_constraint =
+				P::compose_exp(limbs_exp) - reduce_q_exp * p_in_n - P::compose_exp(result_exp);
+			constraints.push(native_constraint);
+
+			constraints.iter().map(|x| s.clone() * x.clone()).collect::<Vec<Expression<N>>>()
+		});
+		selector
+	}
+
+	/// Synthesize the circuit.
+	fn synthesize(
+		self, common: &CommonConfig, selector: &Selector, mut layouter: impl Layouter<N>,
+	) -> Result<Self::Output, Error> {
+		let reduction_witness = self.assigned_integer.integer.reduce();
 		layouter.assign_region(
 			|| "reduce_operation",
 			|mut region: Region<'_, N>| {
-				config.reduce_selector.enable(&mut region, 0)?;
-				Self::assign(&x_limbs, None, &rw, &config, &mut region, 0)
+				selector.enable(&mut region, 1)?;
```

Comment:
Can we use RegionCtx here to enable the selector, and then pass it to `IntegerAssign::assign`.
Apply in all similar cases